python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -1244,6 +1248,175 @@ def knows(self, ref: DatasetRef) -> bool:
+
+        Returns
+        -------
+        existence : `dict` of [`DatasetRef`, `bool`]
+            Mapping from dataset to boolean indicating existence.
+        """
+        # The URIs to be checked and a mapping of those URIs to
+        # the dataset ID.
+        uris_to_check: List[ButlerURI] = []
+        location_map: Dict[ButlerURI, DatasetId] = {}
+
+        location_factory = self.locationFactory
+
+        for ref_id, info in records.items():
+            # Key is the dataId, value is list of StoredItemInfo
+            uris = [info.file_location(location_factory).uri for info in info]
+            uris_to_check.extend(uris)
+            location_map.update({uri: ref_id for uri in uris})
+
+        uri_existence: Dict[ButlerURI, bool] = {}
+        if artifact_existence is not None:
+            # If a URI has already been checked remove it from the list
+            # and immediately add the status to the output dict.
+            filtered_uris_to_check = []
+            for uri in uris_to_check:
+                if uri in artifact_existence:
+                    uri_existence[uri] = artifact_existence[uri]
+                else:
+                    filtered_uris_to_check.append(uri)
+            uris_to_check = filtered_uris_to_check
+
+        # Results.
+        dataset_existence: Dict[DatasetRef, bool] = {}
+
+        uri_existence.update(ButlerURI.mexists(uris_to_check))
+        for uri, exists in uri_existence.items():
+            dataset_id = location_map[uri]
+            ref = id_to_ref[dataset_id]
+
+            # Disassembled composite needs to check all locations.
+            # all_required indicates whether all need to exist or not.
+            if ref in dataset_existence:
+                if all_required:
+                    exists = dataset_existence[ref] and exists
+                else:
+                    exists = dataset_existence[ref] or exists
+            dataset_existence[ref] = exists
+
+        if artifact_existence is not None:
+            artifact_existence.update(uri_existence)
+
+        return dataset_existence
+
+    def mexists(self, refs: Iterable[DatasetRef],
+                artifact_existence: Optional[Dict[ButlerURI, bool]] = None) -> Dict[DatasetRef, bool]:
+        """Check the existence of multiple datasets at once.
+
+        Parameters
+        ----------
+        refs : iterable of `DatasetRef`
+            The datasets to be checked.
+        artifact_existence : `dict` of [`ButlerURI`, `bool`], optional
+            Mapping of datastore artifact to existence. Updated by this
+            method with details of all artifacts tested. Can be `None`
+            if the caller is not interested.
+
+        Returns
+        -------
+        existence : `dict` of [`DatasetRef`, `bool`]
+            Mapping from dataset to boolean indicating existence.
+        """
+        chunk_size = 10_000
+        dataset_existence: Dict[DatasetRef, bool] = {}
+        refs = list(refs)
```

Comment:
I'm a bit confused because I could have sworn that I had already replied to this comment. I can't find it though.

Some of this is history and some of it is the code (not just this bit) being terrified that it might be called with an `Iterator` and not a `Collection`. In this particular case it is probably best to only create new small lists rather than one big list.

There are probably many cases in the code where if we simply assume up top that it was a List then things would be simpler. The problem is that users tend to pass in either an iterator (as returned by `queryDatasets`) or a set (which is the standard advice for what to do with `queryDatasets` output). Iterator isn't Sized but set is. Set is a Collection but you can't index into it.

Reply:
I moved `chunk_iterable` to the utilities package and reimplemented it so it would work without needing `len()` to work. Even though it's only 3 lines long I think it is easier to use as `for chunk in chunk_iterable(x)` format but maybe it's one of those cases where:

```python
for chunk in chunk_iterable(it, 4):
    # use chunk
```
vs
```python
while chunk := tuple(itertools.islice(it, 4)):
    # use chunk
```
is not that much harder to use if you can remember it.